### PR TITLE
Distinguish between truncated and excess content in response

### LINKED ISSUE #3261: Misleading `IncompleteRead` exception, if a HTTP response contains unexpected content
### Subject

When interacting with poorly implemented servers or when paying little attention during mocking of HTTP requests, one can be faced e.g. with a 204 response, which contains an unexpected body. 

So far so bad. urllib3 only raises an `IncompleteRead` exception, which provides no insight into the reason (e.g. a 204 response with body).

### Environment

* Python 3.10
* urllib3 2.1.0

### Steps to Reproduce

````shell
pip install requests responses
````

````python
# scratch.py
import unittest
import responses
import requests


class MyTest(unittest.TestCase):
    url = "http://example.com"

    def test_requestsmock(self):
        rmock = responses.RequestsMock()
        rmock.start()
        rmock.post(url=self.url,
                   status=204,
                   json={"status": "ok"},
                   auto_calculate_content_length=True)

        response = requests.post(url=self.url, data="Foo Bar!")

    @responses.activate
    def test_decorator(self):
        responses.add(method=responses.POST,
                      url=self.url,
                      status=204,
                      json={"status": "ok"},
                      auto_calculate_content_length=True)
        response = requests.post(url=self.url, data="Foo Bar!")


unittest.main()
````

### Expected Behavior

If an empty response body is expected, the raised exception should complain about the presence of an unexpected body.

Also, the reason for expecting an empty body would be helpful and speed up debugging.

### Actual Behavior

Instead, `IncompleteRead(16 bytes read, -16 more expected)` is raised:

```
python3.10 scratch.py
EE
======================================================================
ERROR: test_decorator (__main__.MyTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 712, in _error_catcher
    yield
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 833, in _raw_read
    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
urllib3.exceptions.IncompleteRead: IncompleteRead(16 bytes read, -16 more expected)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/models.py", line 816, in generate
    yield from self.raw.stream(chunk_size, decode_content=True)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 934, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 905, in read
    data = self._raw_read(amt)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 811, in _raw_read
    with self._error_catcher():
  File "/usr/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 729, in _error_catcher
    raise ProtocolError(f"Connection broken: {e!r}", e) from e
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(16 bytes read, -16 more expected)', IncompleteRead(16 bytes read, -16 more expected))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/responses/__init__.py", line 232, in wrapper
    return func(*args, **kwargs)
  File "/home/andi/.config/JetBrains/PyCharm2023.2/scratches/scratch.py", line 28, in test_decorator
    response = requests.post(url=self.url, data="Foo Bar!")
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/api.py", line 115, in post
    return request("post", url, data=data, json=json, **kwargs)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/sessions.py", line 747, in send
    r.content
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/models.py", line 899, in content
    self._content = b"".join(self.iter_content(CONTENT_CHUNK_SIZE)) or b""
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/models.py", line 818, in generate
    raise ChunkedEncodingError(e)
requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(16 bytes read, -16 more expected)', IncompleteRead(16 bytes read, -16 more expected))

======================================================================
ERROR: test_requestsmock (__main__.MyTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 712, in _error_catcher
    yield
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 833, in _raw_read
    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
urllib3.exceptions.IncompleteRead: IncompleteRead(16 bytes read, -16 more expected)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/models.py", line 816, in generate
    yield from self.raw.stream(chunk_size, decode_content=True)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 934, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 905, in read
    data = self._raw_read(amt)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 811, in _raw_read
    with self._error_catcher():
  File "/usr/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/urllib3/response.py", line 729, in _error_catcher
    raise ProtocolError(f"Connection broken: {e!r}", e) from e
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(16 bytes read, -16 more expected)', IncompleteRead(16 bytes read, -16 more expected))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/andi/.config/JetBrains/PyCharm2023.2/scratches/scratch.py", line 18, in test_requestsmock
    response = requests.post(url=self.url, data="Foo Bar!")
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/api.py", line 115, in post
    return request("post", url, data=data, json=json, **kwargs)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/sessions.py", line 747, in send
    r.content
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/models.py", line 899, in content
    self._content = b"".join(self.iter_content(CONTENT_CHUNK_SIZE)) or b""
  File "/home/andi/virtualenvs/checkers/lib/python3.10/site-packages/requests/models.py", line 818, in generate
    raise ChunkedEncodingError(e)
requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(16 bytes read, -16 more expected)', IncompleteRead(16 bytes read, -16 more expected))

----------------------------------------------------------------------
Ran 2 tests in 0.005s

FAILED (errors=2)
```

### PR DESCRIPTION
**Pull request by crazyscientist**
`HTTPResponse._raw_read` raises `IncompleteRead` if the content length does not match the expected content length. For malformed responses (e.g. 204 response with content) the re-raised `ProtocolError` was a bit too unclear about the unexpected excess content being the reason for the exception.

With this change, the exception points out, that the client is not dealing with a connection error but a protocol violation.

Closes #3261

### CODE REVIEW

**Review by illia-v — APPROVED** [review]
LGTM, thanks!

**Review by sethmlarson — COMMENTED** [review]
Some comments to not hold up review from others:

File: src/urllib3/response.py
Comment (illia-v) [reply src/urllib3/response.py:755]: It's integer https://urllib3.readthedocs.io/en/stable/reference/urllib3.exceptions.html#urllib3.exceptions.IncompleteRead

Comment (crazyscientist) [reply src/urllib3/response.py:755]: I have noticed the int type annotation in the signature of `IncompleteRead.__init__`. But the `nox` linter assumes that it's bytes; without the explicit type cast I get the following complaint:

```
src/urllib3/response.py:755: error: Non-overlapping equality check (left operand type: "int", right operand type: "bytes")  [comparison-overlap]
```

Please advise me on your preferred strategy to address this issue.

Comment (illia-v) [reply src/urllib3/response.py:755]: I'd define types of `IncompleteRead` attributes this way:
```python
class IncompleteRead(HTTPError, httplib_IncompleteRead):
    partial: int  # type: ignore[assignment]
    expected: int
```

This leads to removing a few `type: ignore`, and then works.

Comment (crazyscientist) [reply src/urllib3/response.py:755]: Makes perfect sense to me :slightly_smiling_face: 

I had considered that, but it lead to a few test failures, too. So, I went for the explicit type conversion.

But, as you like that idea, let's see where it takes us :+1:

File: src/urllib3/response.py
Comment (crazyscientist) [reply src/urllib3/response.py:757]: I tried to come up with something that followed a similar structure as the other messages.

What do you think about this version?

```suggestion
                    arg = f"Response may not contain a body: {e!r}"
```

Comment (illia-v) [reply src/urllib3/response.py:757]: > What do you think about this version?

Looks good to me.

Maybe we should drop `: {e!r}` too.
In the traceback from https://github.com/urllib3/urllib3/issues/3261#issue-2062496957 the error was `urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(16 bytes read, -16 more expected)', IncompleteRead(16 bytes read, -16 more expected))`. I don't see why we need that `IncompleteRead(16 bytes read, -16 more expected))` duplicated.

Comment (illia-v) [reply src/urllib3/response.py:757]: I meant dropping `: {e!r}` just for the new `Response may not contain a body`

File: src/urllib3/response.py
Comment (crazyscientist) [reply src/urllib3/response.py]: My apologies, took me some time to get back to this... Please see e08676ee

**Comment by ecerulm** [comment]
@crazyscientist ,can you update this branch to the latest main?

**Review by ecerulm — CHANGES_REQUESTED** [review]
Please remove the "Protocol violation:", update the branch to latest main, and change `urllib3.exceptions.IncompleteRead` use the type hints in the way that @illia-v  suggested.

File: src/urllib3/response.py
Comment (crazyscientist) [reply src/urllib3/response.py]: >  to redefine `urllib3.exceptions.IncompleteRead.partial`

@ecerulm how does your suggestion differ from my change I committed in 12ce2a76a927a4ed28760baea878e5f69eb91e4b?

> `nox -rs text-3.12` only fail on `test_unexpected_body`

Where is the session "text-3.12" defined? After rebasing I keep getting this error:

```
$ nox -rs text-3.12
nox > Error while collecting sessions.
nox > Sessions not found: text-3.12
$ python3 --version
Python 3.12.1
```

File: src/urllib3/response.py
Comment (crazyscientist) [reply src/urllib3/response.py]: Changed in b16536a1

File: src/urllib3/response.py
Comment (ecerulm) [reply src/urllib3/response.py]: > @ecerulm how does your suggestion differ from my change I committed in 12ce2a7?

The type hints are outside of `__init__`. Full source code


```python
class IncompleteRead(HTTPError, httplib_IncompleteRead):
    """
    Response length doesn't match expected Content-Length

    Subclass of :class:`http.client.IncompleteRead` to allow int value
    for ``partial`` to avoid creating large objects on streamed reads.
    """

    partial: int  # type: ignore[assignment]
    expected: int

    def __init__(self, partial: int, expected: int) -> None:
        self.partial = partial
        self.expected = expected

    def __repr__(self) -> str:
        return "IncompleteRead(%i bytes read, %i more expected)" % (
            self.partial,
            self.expected,
        )
```

But , if you asking what it's the difference **functionally** then I believe they are equivalent although I think this style is preferred. I'll let @illia-v and @sethmlarson  decide on that (since I'm fairly new here), but at least is the one that is used in `urllib3.connectionpool.ConnectionPool`, etc.

Comment (crazyscientist) [reply src/urllib3/response.py]: @ecerulm Yes, being new to this code base, it did not occur to me to think of applying the type annotation this way, because I too believe, that both versions yield the same result.

It's your project with your design guidelines. Let's follow that one. Please see c3cdaed0

**Review by illia-v — APPROVED** [review]
LGTM, thanks!

**Review by ecerulm — APPROVED** [review]

**Pull request merged**
